### ISSUE TITLE
Incorrect LAN broadcast address

### ISSUE THREAD

**Issue by hiker** [supertuxkart/stk-code#3187]
When trying to discover LAN servers, the broadcast address 255.255.255.255 is used. But this does not (always?) work, on certain networks the broadcast is not delivered (or over 50 seconds to arrive).

But it appears that at least in some cases the broadcast works if a 'proper' broadcast address like 192.168.0.255, but I seem to have problems when using 192.168.255.255.

We could either try to broadcast using various addresses (and remove duplicated responses). or see if there is a way to find the right netmask so we can compute the right broadcast address.

**Comment by hiker** [comment]
Deve's problems are fixed after a reboot of the router, and he can now connect with 255^4, 192.168.1.255.

255^4 apparently is known to have issues on windows(only??), so best option would be to find the right broadcast address.

With my pi things are a bit strange: 255^4 does not work. 192.168.0.255 works (as does using the actual IP address), but 192.168.255.255 does NOT work. The latter is strange since the network should be a /16, which is also what windows GetAdaptersAddresses reports.

But in my router config I see a LAN netmask of 255.255.255.0, and a WAN mask of 255.255.248.0.
Since I don't think I can query the router data, the only option I see is to add code that determines the netmask (which windows reports to be 255.255.0.0 for me) - but then loop over all possible options, i.e. for me not only using 0.0.255.255 as or-ed broadcast address, but also 0.0.0.255 (since this is the one that works for me).

This means that a client might receive several responses from the same server - so we need to remove duplicate servers on the client (trying to only send out one response to a query on the server is probably too difficult, since clients might refresh frequently, and we might have more than one client running on a system (at least for development/debugging).
@auriamg, @deveee, @Benau - any better ideas?

**Comment by Benau** [comment]
I suggest u add connect-now equivalent to stk network gui

**Comment by hiker** [comment]
I have already agreed with you and @deveee that we need to add a gui - this ticket is about fixing the automatic detection, broadcast to 255^4 as we do atm is not correct. 

I have started working on the windows side.

**Comment by deveee** [comment]
Actually it wasn't reboot that helped, because I did reboot before. I had to change some settings and restore it back, so I'm not really sure what happened :P

And I think that ~30-60 seconds is acceptable to discover IP address. So we can use different methods to get the IP or we can even manually send packets to all local addresses if broadcast packet has no answer. 

And when server IP is found, then we can store it in config.xml, because it is probable that it will be the same address next time.

And +100 for connect-now in stk gui. At least a small button somewhere.

**Comment by deveee** [comment]
Btw. it seems that in global network there is still active "devetest's server" that I created yesterday. It was probably hard close with Ctrl+C but should it automatically clear it after some time?

**Comment by Benau** [comment]
it can only be automatically cleared after someone create another new server, server has no "main loop" concept

**Comment by deveee** [comment]
Ok, I just wasn't sure what I should expect.

What do you think about port detection in connect-now dialog? Because it's just standard connection with known IP and 2757 port, so if it works, then there is no need to check/write real port number. And user still can write full IP address with port number. Here is a quick patch:
```
diff --git a/src/network/servers_manager.cpp b/src/network/servers_manager.cpp
index e07e98c23..4e97e8ff6 100644
--- a/src/network/servers_manager.cpp
+++ b/src/network/servers_manager.cpp
@@ -197,6 +197,68 @@ Online::XMLRequest* ServersManager::getLANRefreshRequest() const
 
 }   // getLANRefreshRequest
 
+
+
+unsigned int ServersManager::detectPort(uint32_t requested_ip)
+{
+    unsigned int found_port = 0;
+    
+    ENetAddress addr;
+    addr.host = STKHost::HOST_ANY;
+    addr.port = STKHost::PORT_ANY;
+    Network *network = new Network(1, 1, 0, 0, &addr);
+
+    BareNetworkString s(std::string("stk-server"));
+    TransportAddress address(requested_ip, 
+                       NetworkConfig::get()->getServerDiscoveryPort());
+    network->sendRawPacket(s, address);
+
+    Log::info("ServersManager", "Detect port for IP address.");
+
+    const int LEN=2048;
+    char buffer[LEN];
+
+    double start_time = StkTime::getRealTime();
+    const double DURATION = 1.0;
+
+    while (StkTime::getRealTime() - start_time < DURATION)
+    {
+        TransportAddress sender;
+        int len = network->receiveRawPacket(buffer, LEN, &sender, 1);
+        
+        if (len > 0)
+        {
+            BareNetworkString s(buffer, len);
+            uint8_t version = s.getUInt8();
+            
+            if (version != NetworkConfig::m_server_version)
+            {
+                Log::verbose("ServersManager", "Wrong server version.");
+                continue;
+            }
+            
+            irr::core::stringw name;
+            // bytes_read is the number of bytes read
+            s.decodeStringW(&name);
+            uint8_t max_players = s.getUInt8();
+            uint8_t players     = s.getUInt8();
+            uint16_t port       = s.getUInt16();
+            uint8_t difficulty  = s.getUInt8();
+            uint8_t mode        = s.getUInt8();
+            uint8_t password    = s.getUInt8();
+            
+            found_port = port;
+
+        }   // if received_data
+    }    // while still waiting
+
+    delete network;
+    
+    return found_port;
+}   // operation
+
+
+
 // ----------------------------------------------------------------------------
 /** Factory function to create either a LAN or a WAN update-of-server
  *  requests. The current list of servers is also cleared.
diff --git a/src/network/servers_manager.hpp b/src/network/servers_manager.hpp
index 823ce74b3..86a247d99 100644
--- a/src/network/servers_manager.hpp
+++ b/src/network/servers_manager.hpp
@@ -72,6 +72,8 @@ public:
     std::vector<std::shared_ptr<Server> >& getServers()   { return m_servers; }
     // ------------------------------------------------------------------------
     bool listUpdated() const                         { return m_list_updated; }
+    // ------------------------------------------------------------------------
+    unsigned int detectPort(uint32_t requested_address);
 
 };   // class ServersManager
 #endif // HEADER_SERVERS_MANAGER_HPP
diff --git a/src/states_screens/online_screen.cpp b/src/states_screens/online_screen.cpp
index 04ec57e17..57c1dc0bd 100644
--- a/src/states_screens/online_screen.cpp
+++ b/src/states_screens/online_screen.cpp
@@ -31,6 +31,7 @@
 #include "network/protocols/client_lobby.hpp"
 #include "network/network_config.hpp"
 #include "network/server.hpp"
+#include "network/servers_manager.hpp"
 #include "network/stk_host.hpp"
 #include "online/request_manager.hpp"
 #include "states_screens/networking_lobby.hpp"
@@ -242,8 +243,20 @@ void OnlineScreen::eventCallback(Widget* widget, const std::string& name,
             {
                 TransportAddress server_addr(
                     StringUtils::wideToUtf8(tb->getText()));
+                    
+                if (server_addr.getIP() != 0 && server_addr.getPort() == 0)
+                {
+                    unsigned int port = ServersManager::get()->detectPort(server_addr.getIP());
+                    
+                    if (port != 0)
+                    {
+                        server_addr.setPort(port);
+                    }
+                }
+                
                 if (server_addr.isUnset())
                 {
+
                     core::stringw err = _("Invalid server address: %s.",
                         tb->getText());
                     lw->setText(err, true);
```

**Comment by Benau** [comment]
I have granted you a wish b8e7701bcb8035b417cd4fc8360ec3fbd3f5c10e

**Comment by hiker** [comment]
@nado Well, what I meant: While 192.168 is a /16, my router only supports it as /24 (I can only set the last byte of the netmask). Not sure where windows its getting its netmask from. I am just finishing some other code, then I'll come back to this and push my changes

@leyyin, @deveee I agree, reading /proc feels indeed wrong.

**Comment by nado** [comment]
It depends in what the active dhcp server gives, and if you have several, I’d suggest ignoring it as this is not advised and most people won't encounter this situation. You could try setting your network configuration manually to work around it.

+1 for getifaddrs

**Comment by deveee** [comment]
Actually ifconfig reads /proc/net/dev too. And it's just to get interface names. And I remember that I had problems with getifaddrs but I don't remember why :P Maybe it will work fine.

**Comment by hiker** [comment]
I've tried to fix those issues on:
https://github.com/supertuxkart/stk-code/tree/lan-broadcast
Seems to work fine under Windows, I couldn't really test linux, so this still needs to be tested and potentially fixed.

Two main questions:
1) is adding all the various broadcast addresses (e.g.*.*.255.255, *.*.*.255 etc) useful?
2) Should we also add 255.255.255.255?
3) Any comments re the list of default LAN broadcast addresses?

**Comment by hiker** [comment]
If you have multiple interfaces I don't think there is anything we can do (short of displaying the IP addresses with each server to allow picking the right one; or maybe running a few pings to find the fastest one). Originally I used the IP address of the server as key in the mapping (which automatically removed duplicates), but then had the same server listed twice (once as 192.*, once as 127.*).

And yes, I do get the virtual interface as well, but since there is no STK instance running there, that should not matter - there is no server to answer from there.

TBH, I am happy to ignore those issues for now. If we get real cases in which we have a problem, we can tweak this. We do have access to the interface name, e.g.:
```
Interface: lo   Address: 127.0.0.1      mask: 28b4f4c -> ffffff
Interface: em1  Address: 134.178.30.198 mask: 28b5004 -> 1ff
```
(no VM on my linux machine though). I am not sure if this really helps us - I guess we could try to collect a set of interface names to be ignored (but e.g. what if we want to run a server on a VM to test a particular OS?).

I've added 192.168.1.255 to the list (and removed duplicated entries ;) ).

So I think for now we can close this ticket - hooray ;)
